### DISCUSSION TITLE
The freedom and increased limits of object selection are incredible...The handling of scenery tabs/groups is not and it is a direct hindrance to the advancement in object selection you blessed us with!

### DISCUSSION THREAD

**Discussion by FischNE** [OpenRCT2/OpenRCT2#19892] [Feature requests]
Hello everyone, Fisch here from NE. First of all, thanks once more for everything. The advancements this project has brought for RCT are truly unbelievable. You guys rock!! Now to jump in: I have a request that in my opinion would increase the quality, efficiency, and fun of scenery oriented sandbox building by a huge amount.

Thanks to you, we have the option to select way more objects than the game originally allowed us to. This gives us great freedom when building because we can A) place more map data, B) use more different objects to start with and C) select additional objects for the map at any time. When you want to build something and you realize you don't have the right objects for it yet, you can just go into object selection and select some other objects that allow you to better get the desired visuals. This is absolutely incredible for RCT and it has already led and will further lead to unbelievable sandbox maps.
That said, due to this advancement, the scenery window in the game is just not up to par anymore and I will explain why.

RCT originally came with vanilla scenery groups such as "Roman Scenery", which probably brought us around 30-45 Roman themed objects. For a scenery window which in total only allowed us to select up to 256 small scenery objects, that is an ok size for a scenery group/category. Now though, we have more than 2000 object selections available in small scenery alone. I have a map where I have selected roughly 1300 small scenery objects. I have downloaded probably thousands of parks on NE over the past 1.5 decades, so I have a huge amount of custom scenery groups. I have created scenery groups myself. I have created sandbox bench maps myself. Yet, I simply cannot say that I have a single bench map which would fully categorize 1500+ small scenery objects all into well ordered and thematically correct groups. This is especially true when rare miscellaneous objects are included that are probably not part of any existing scenery groups at all. Instead my park files' scenery groups usually look somewhat like this now:
As an example the scenery group I have for trees will have around 50 trees. The miscellaneous "?" scenery group will have another 50+ trees scattered among random other things. Overviewing all the selected objects for the desired type of scenery becomes almost impossible.

Anyway though, I actually am among the <1% of people who often have manually created new scenery groups for the hundreds (now thousands) of objects I have selected at the start of a new project. Still that alone would not solve the issue for my project. Throughout the duration of the project, I will select additional objects that will then not be included in the custom scenery groups that were previously created specifically for this project.
The part of the community that begged for the object selection limits to be loosened in the first place, has continuously made advances in object creation over the past 2 decades. Nowadays it is completely normal to manually draw new objects for your bigger projects yourself. I have created hundreds of scenery objects over the years. People draw new objects while working on their projects. So even if someone had gone the distance of creating their own scenery groups for the thousands of objects that they had selected at the start, within a very short amount of time they will have added new objects to the object selection that will then end up in the "?" scenery group. This will be the case for hundreds of objects over the course of a project. My "?" scenery group currently usually holds more plant objects, more tree objects, more wall objects, more window objects, and more miscellaneous objects than all the groups I had specifically created for these purposes combined. Add on top of that, that as the project progresses, some objects might become more important than others. In theory I would want to readjust the order of my scenery groups, so that the most important objects are further to the top than the least important ones.

To sum this up, we are able to select objects at will for them to be available as we play and build, but we are not able to then ever find them in the scenery window because we have so many objects available all while we cannot get them to be in the correct groups and to be ordered correctly. Searching for the object name is also not a solution because of all the obscure names that custom objects have and because nobody will ever memorize the names of all the objects that were selected.

The biggest quality of life improvement for sandbox RCT that I could think of at this current point would be **improved, flexible, and rearrangeable scenery groups**. This is my first priority not because I'm not eager for other potential new features. I am eager for anything you guys come up with. But quite simply my head explodes every time I try to find the right objects to build. I want to build a complex structure with custom scenery but I spend probably 75+% of the time just browsing through the scenery window in order to assess what objects I have available for my task. That's for me, a player that has been building with custom scenery since 2005. I can't imagine that others are finding this too easy either.

SORRY for this wall of text but I think explaining why this is even a big thing is very important, as it'd be such a big task.
In the past year I had thought, maybe this is something to be addressed with a tool outside the game. I had contacted Sadret, the creator of the scenery manager plugin as well as Trigger, the creator of Trigger's Group Creator. I had also proposed this on multiple discords. Through all the discussions that came out of this though, I think I have come to the conlusion, that clearly the best solution from a player perspective would be a solution within OpenRCT.

Features I am thinking of for a tool that would be a solution to this:

- _ability to create a new scenery group, give it a name and a preview image_
- **ability to add objects to scenery groups on the fly by dragging/dropping them from a list of preview images of all selected objects on a given map, as well as from an "add file" window**
- **ability to delete objects from said scenery group while looking at preview images of the included objects**
- **ability to reorder scenery groups on the fly by dragging/dropping the objects into different positions within that scenery group**
- **ability to select multiple objects for the above actions at once, so the tasks would be much faster**
- ability to undo/redo the last change you made while creating that scenery group
- ...

For a potential external program I had come up with the following image as a sketch for the UI in the past:
![Scenery Group Creator](https://user-images.githubusercontent.com/18613267/230799803-7749f31c-68f3-4cff-93a9-16d10f90d99e.png)

Thanks to anyone that just took the time to read all of this. If something comes out of this I would be absolutely delighted. I totally understand, that it'd be a huge task. In terms of RCT it would be an incredible quality of life improvement however!

Wish you guys all the best
Fisch

**Comment by ocalhoun6** [comment]
Sounds like a job for a plugin... "Scenery Group Manager" or something like that.

**Comment by X123M3-256** [comment]
> I think I have come to the conlusion, that clearly the best solution from a player perspective would be a solution within OpenRCT.

Is it, though? It sounds like what you want is essentially a scenery group editor tool ... and while that could be built into the game, does it really need to be? Object creation tools are generally separate from the game itself, at least thus far. The only thing we do have built in is the sprite compiler, I think.

I don't think you can hot reload objects currently which means you would need to reload the game after editing an object for it the changes to show up ... which I think would eliminate the advantage of having it built into the game. I'm not sure how hard it would be to make it so objects can be reloaded on the fly, though, maybe that's doable.

It does to me sound like something like this would really be best as a plugin, but also, plugins don't have access to the filesystem so I'm not sure that's possible right now.  If you're creating new scenery tabs specifically for one park, though, I wonder if a better solution wouldnt be an option to have a plugin that acts as an alternate scenery window, ignoring the loaded scenery group objects and storing its own list of scenery tabs in the park file that applies only to that park file.

Actually, one could maybe argue that this should be how it works by default - perhaps scenery group objects could be essentially done away with and used only to initialize a list of tabs that is local to the park file. This would allow the grouping for a given park to be freely edited in an ad-hoc manner, perhaps you could simply drag and drop scenery objects from one tab to another within the scenery window itself. But while this sounds like it might be ideal for your use case, I'm not sure if it would be a good idea generally and it seems odd to have objects that are never actually loaded (but also you can't just drop them entirely because existing parks use them)

If think if the underlying object is edited then the interface needs to make that clear - otherwise people might edit default tabs without realizing the changes will apply globally, or we'll end up with multiple identically named but conflicting scenery tabs.

**Comment by SixFrags** [comment]
While a solution might be hard, as X123M3-256 explained hot (re)loading objects isn't possible currently, maybe a plugin that could sort and group scenery objects could be a solution? 
Unfortunately I'm not capable of creating something like this myself and I just hope someone would be able and willing to implement a feature like this, as it's the most important feature the game is missing at the moment (at least for me personally).

**Comment by Sadret** [comment]
Does the plugin API have the capabilities needed for implementing an alternatie scenery window? As far as I know, there is currently no way to get the image id for some scenery element. But I guess that should be decently easy to add.

For the sorting/managing of existing scenery groups, I think it might be too much work to add this to the API.

**Reply by IntelOrca** [reply]
https://github.com/OpenRCT2/OpenRCT2/pull/17567

**Reply by Sadret** [reply]
PR description looks good, merging this will definitely help to develop a plugin that offers a custom scenery window.

**Reply by FischNE** [reply]
This sounds very promising!

**Comment by FischNE** [comment]
I agree with the statement above that for an ingame solution perhaps a plugin that pretty much acts as an alternative scenery window would be the way to go. Maybe this will allow for a lot of freedoms and potential going forward.
I'm very happy and thankful in regards to the replies and any activity related to this. It would be a huge quality of life improvement for custom scenery oriented playstyles in OpenRCT2.

**Comment by duncanspumpkin** [comment]
I think a plugin solution would be the only way to fully satisfy all NE types but we could still look at a few other nice to haves for the default window. We could add a "star" button for all scenery when starred they would be added to a star tab. This could be dynamically done and saved to the park file. You see this feature on almost all photo apps so users will immediately understand it. @kyphii suggested a recently used tab as well. Both of these are pretty easy to understand and use and would likely aid the majority of users.

**Reply by ocalhoun6** [reply]
> @kyphii suggested a recently used tab as well. 

That would actually be incredibly useful, even for very casual players who aren't doing super-intensive scenery work with thousands of scenery items.

There are lots of times when you're working on some specific batch of scenery when you might need to switch back and forth frequently between two or more different scenery items. (For example, if you're making a garden maze with flowerbeds within fences, you'd need to frequently switch back and forth between flowerbed scenery and fence scenery.) If there was a 'recently used' tab, then once you'd placed at least one each of the scenery items you're working with, you could just stay on the recently used tab and easily be able to switch items with a single click (as opposed to an extra click for switching tabs and maybe also some scrolling to find the item within that tab).

It could also help you make things uniform, because you never have to ask yourself, "Wait ... which kind of tree did I use on the other side of the park again? Am I selecting the correctly matching one, or just one that looks similar, but not quite the same?" Because you could always go to your 'recently used' tab and find _only_ the items you've just used -- then you can be sure they'll match.

Hopefully the recently used list will be saved with the scenario, though. That way, you don't have to worry about losing that progress if you have to take a break from playing. (Need to make sure the list is cleared when creating a new scenario, though!)

**Reply by duncanspumpkin** [reply]
Well do keep in mind the pipette tool can be used for that as well

**Reply by ocalhoun6** [reply]
Well, yeah ... but that's still an extra click.

With a 'recently used' tab, you could select recent scenery with only a single click directly on the scenery item in the tab. But the pipette tool requires two clicks to select the scenery: one click on the tool to activate it, then another click to select the item.

One extra click doesn't sound like much, but if you're switching back and forth very frequently over the course of a large project, it could add up to a lot of extra clicks.

**Comment by Sadret** [comment]
A problem with the plugin approach would be missing functionality: quarter tile placement, ctrl / shift modifier
